write CF/UGRID output

dabele opened the pull request:
This adds the ability to output NetCDF files in [CF](https://cfconventions.org)/[UGRID](http://ugrid-conventions.github.io) format, as discussed in #18903

Usage is similar to hdf5, using `DataOutFilter` with added support for custom attributes and time series.

I think I added a reasonable amount of tests (more than hdf5 anyway :) ) I didn't want to write a full copy of `ncdump`.

I also tested the output in both ParaView and cf-plot. Both could display the files written by deal.ii. cf-plot doesn't handle discontinuous output well, but that's expected as it mostly does contour plots, which don't really make sense with discontinuous fields anyway.

Added NetCDF as an optional dependency.

Limitations:
- Only 2D. 3D unstructured grids are part of UGRID conventions, but have not been added to CF yet. The proposed format for 3D is also a bit different from 3D, so I decided to skip it for now.
- No mixed meshes. Supported by CF, but not by `DataOutFilter` I think. The cell node indices are all in one vector. I guess DataOutFilter could be extended to support this somehow. But grids only of either only triangles or only quads work.
- No vector valued data, not supported by CF, must be split into its components and given unique names. I guess we could split the components and generate the names automatically?

drwells commented on include/deal.II/base/exception_macros.h:
```diff
@@ -1049,4 +1049,27 @@
+ * @ingroup Exceptions
+ */
+#  define AssertThrowNC(code) \
+    AssertThrow(code == NC_NOERR, ExcIO(nc_strerror(code)))
```
This assumes that we have `nc_strerror()` and `NC_NOERR` in the current namespace, which might not be the case. Could you make this work without requiring those to be in this macro?

For some related libraries (like ExodusII) we hard-code the value of `EX_NOERR`.

bangerth replied:
I think a related problem is that people may be using this macro outside `namespace dealii`, so you need to say `::dealii::ExcIO` instead. And you should use `::nc_strerror` in case this macro is used in a namespace that has its own `nc_strerror` symbol.

Separately, is `ExcIO` the right error? Can these errors only be I/O errors, or other things as well? I think `ExcMessage` might be the better choice.

dabele replied:
This is also modelled after the hdf5 code which uses ExcIO.

I think actual netcdf errors can only be a) IO errors or b) bugs in the implementation. All other errors should be excluded by the data structures (`patches`, `CFFlags`, etc.) or dedicated checks (e.g. no mixed meshes, wrong dimension, etc). The dedicated checks should not use ExcIO, they currently are, but I will change that. I feel ExcIO is correct for the netcdf macro since IO errors (e.g. file opened somewhere else, or existing file with invalid format) is the only type of netcdf error that a user could handle, everything else is an internal bug.

You are right about the namespace issue.

drwells commented on source/base/data_out_base.cc:
```diff
@@ -8797,8 +8817,562 @@ DataOutBase::write_hdf5_parallel(
+namespace
+{
+  /**
+   * Automatically close the netcdf file on scope exit.
```
We can use our own `ScopeExit` class for this

dabele replied:
I considered this, but I think a separate class is warranted, the class has expanded now to fully manage opening and closing of netcdf files with more error checking. this is safer and less repetition.

drwells commented on source/base/data_out_base.cc:
```diff
@@ -8797,8 +8817,562 @@ DataOutBase::write_hdf5_parallel(
+                const DataOutBase::CFFlags       &flags,
+                const std::string                &filename,
+                const int                         global_node_cell_count[2],
+                const int                         global_node_cell_offsets[2],
```
This isn't standard syntax (I'm surprised we don't get warnings on the CI): could you use `std::array` instead?

dabele replied:
created a utility struct and function to synchronize the local and global mesh sizes.

Also used it in the hdf5 functions.

drwells commented on source/base/data_out_base.cc:
```diff
@@ -8797,8 +8817,562 @@ DataOutBase::write_hdf5_parallel(
+                         return patch.reference_cell == ref;
+                       }),
+           ExcNotImplemented(
+             "Mixed meshes are currently not supported in CF output."));
```
True, but you can get this information directly from `Triangulation()` via `is_mixed_mesh()`.

dabele replied:
The problem is that Triangulation is not available at this point, only the patches are (through `DataOutInterface`)

drwells reviewed (dismissed):
I apologize for the long delay in posting my review. There's a lot of good stuff here: I'm happy to see more binary file formats supported.

This is a pretty large PR. Would you be willing to start by moving the basic NETCDF (i.e., detection via CMake and CI stuff) into another PR?

dabele commented:
Thank you for your first comments

As you suggested, I created #19296 to isolate the cmake stuff, I already applied some of your suggestions and I answered all the build related comments there so the discussion here can focus on the rest of the code.

dabele committed:
Add NetCDF output with CF conventions

dabele commented on cmake/modules/FindDEAL_II_NETCDF.cmake:
```diff
@@ -0,0 +1,53 @@
+    COMMAND ${DEAL_II_NETCDF_CONFIG_PROGRAM} "--has-nc4" 
+    OUTPUT_VARIABLE _has_nc4)
+  if(NOT _has_nc4)
+    message(WARNING "Ignoring the discovered NetCDF library because it does not support format version 4.")
```
covered in #19296

dabele commented on cmake/modules/FindDEAL_II_NETCDF.cmake:
```diff
@@ -0,0 +1,53 @@
+  if(NOT _has_nc4)
+    message(WARNING "Ignoring the discovered NetCDF library because it does not support format version 4.")
+  else()
+    execute_process(
```
covered in #19296

dabele commented on cmake/configure/configure_50_netcdf.cmake:
```diff
@@ -0,0 +1,30 @@
+  endif()
+endmacro()
+
+configure_feature(NETCDF)
```
covered in #19296

dabele commented on source/base/data_out_base.cc:
```diff
@@ -8797,8 +8817,562 @@ DataOutBase::write_hdf5_parallel(
+      // Ignore the error code because nc_close can only fail
+      // if ncid is invalid, in which case nc_create/open or
+      // other functions already failed.
+      nc_close(ncid);
```
Done. I expanded the RAII class that holds and closes the netcdf file to avoid repeating this.

bangerth commented on include/deal.II/base/data_out_base.h:
```diff
@@ -1084,6 +1085,107 @@ namespace DataOutBase
+     * The map may include the special keys that do not correspond to
+     * data vectors (and should not be used as names for data vectors):
+     * - "global": define global attributes that describe the file content
+     * - "time", "x", "y": attributes for dimension variables, e.g., "units".
```
How about "z"?

dabele replied:
Certainly should be added if there is 3D, see below.

bangerth commented on include/deal.II/base/data_out_base.h:
```diff
@@ -1084,6 +1085,107 @@ namespace DataOutBase
+     * data vectors (and should not be used as names for data vectors):
+     * - "global": define global attributes that describe the file content
+     * - "time", "x", "y": attributes for dimension variables, e.g., "units".
+     * Attributes are only written when the variable is first created, not
```
did you mean *file* instead of *variable* here?

dabele replied:
I meant variable, but the variable is created at the same time as the file, and file is probably easier to understand. I just mean that you can't add attributes on the second time step. Obviously not a strict limitation, just the way it is implemented right now.

bangerth replied:
I see. We're big fans of extensive documentation, so whenever you find that you writing a sentence or three of explanation in response to a review comments, just put those into the documentation of the code as well.

dabele replied:
added

bangerth commented on include/deal.II/base/data_out_base.h:
```diff
@@ -1084,6 +1085,107 @@ namespace DataOutBase
+     * - "global": define global attributes that describe the file content
+     * - "time", "x", "y": attributes for dimension variables, e.g., "units".
+     * Attributes are only written when the variable is first created, not
+     * on subsequent calls with \ref keep_existing_file = true;
```
Here and elsewhere, please use `@ref` instead of `\ref` -- the former is the form we use everywhere else.

dabele replied:
Done.

bangerth commented on include/deal.II/base/data_out_base.h:
```diff
@@ -2957,6 +3076,27 @@ class DataOutInterface
+   * @code
+   * // Filter the data and store it in data_filter
+   * DataOutBase::DataOutFilterFlags flags(true, true);
+   * DataOutBase::DataOutFilter data_filter(flags);
```
Here and in the example above, you're using `DataOutFilter`, which is something we don't usually do. Is this necessary? It means that you can't output discontinuous quantities, for example. In any case, it might be nice to have two sentences here and above where you briefly explain why `DataOutFilter` is necessary, if that is indeed the case.

dabele replied:
I think you can still have discontinuous quantities if you don't enable the vertex filter. I will be using discontinuous quantities a lot, so this is important to me.

Using the DataOutFilter is not necessary of course. I modelled the netcdf code on the hdf5 code which I believe also only works with DataOutFilter. This is partly because they seem to serve a similar purpose, but also because the implementation started out as an extension outside of dealii and this was the easiest way, the class does all of the preprocessing already. Climate data sets are also often large, so having a filter seems useful? Not sure what the best way is.

I'm open to changing it. This would basically mean implementing something like `VtkStream` for netcdf, right?

bangerth replied:
Actually, I don't mind if you want to leave the example code snippets this way -- that's fine, just describe in the surrounding documentation why you're doing it that way and what consequences this has. For example, you could say here that the use of the first `true` argument to `DataOutFilterFlags` implies that data at the same vertex obtained from adjacent cells will all be unified, and will consequently result in a continuous field being saved, even if the original finite element field was discontinuous. And then say what you'd have to change to get that field to show up as a discontinuous field.

(Ideally, there would then of course also be a test for each of the two ways.)

dabele replied:
Sorry, I thought you were asking why it's implemented this way.

The examples use `DataOutFilter` because that's currently the only way to use the CF output feature. `DataOutFilter` is well documented regarding continuous/discontinuous output, I feel that shouldn't be repeated here. But I added a short note and reference to `write_cf_parallel`, that would likely be the first contact for users with the new feature.

There is a test using both filtered and unfiltered vertices, since there are multiple tests anyway and it is easy to set the flag to true in one or more of the tests. But I feel this is really testing `DataOutFilter` and not CF output, nothing changes in the logic of `write_cf_parallel` when the vertices are filtered or not.

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8245,6 +8265,61 @@ DataOutInterface<dim, spacedim>::write_filtered_data(
+                        unsigned int   num_cells_local,
+                        const MPI_Comm comm)
+  {
+    unsigned int num_local[]{num_nodes_local, num_cells_local};
```
Like in the lines below, let's give this object an explicit dimension:

```suggestion
    unsigned int num_local[2] = {num_nodes_local, num_cells_local};
```

dabele replied:
Done

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8245,6 +8265,61 @@ DataOutInterface<dim, spacedim>::write_filtered_data(
+   * Synchronize numbers of cells and nodes of a distributed mesh.
+   */
+  DistributedMeshSizes
+  synchronize_mesh_size(unsigned int   num_nodes_local,
```
This isn't quite the right function name since the function is not synchronizing the contents of variables it has access to and that live in different spaces. Rather, it computes something and returns it. So perhaps call it `compute_global_mesh_size()` or something like this?

dabele replied:
Done

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8245,6 +8265,61 @@ DataOutInterface<dim, spacedim>::write_filtered_data(
+   */
+  DistributedMeshSizes
+  synchronize_mesh_size(unsigned int   num_nodes_local,
+                        unsigned int   num_cells_local,
```
These two are input arguments, so let's mark them as constant:

```suggestion
   * Synchronize numbers of cells and nodes of a distributed mesh.
   */
  DistributedMeshSizes
  synchronize_mesh_size(const unsigned int   num_nodes_local,
```
Same for the other function argument.

dabele replied:
Done

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8245,6 +8265,61 @@ DataOutInterface<dim, spacedim>::write_filtered_data(
+    unsigned int num_nodes_local;
+    unsigned int num_cells_local;
+    unsigned int num_nodes_global;
+    unsigned int num_cells_global;
```
We typically call these variables `n_something`. Would you mind changing the names?

dabele replied:
Done

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8267,16 +8342,10 @@ namespace
     hid_t pt_data_dataspace, pt_data_dataset, pt_data_file_dataspace,
       pt_data_memory_dataspace;
     herr_t              status;
-    std::uint64_t       local_node_cell_count[2];
```
The old code used `std::uint64_t` for these variables, but your new code uses `unsigned int`. I think the latter is not big enough -- we sometimes have simulations with more than 4B cells and nodes. Can you turn this back into `std::uint64_t`?

dabele replied:
Done.

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8295,31 +8364,7 @@ namespace
-    global_node_cell_count[1]   = local_node_cell_count[1];
-    global_node_cell_offsets[0] = global_node_cell_offsets[1] = 0;
-#  endif
+    auto mesh_sizes = synchronize_mesh_size(data_filter.n_nodes(), data_filter.n_cells(), comm);
```
```suggestion
    const auto mesh_sizes = synchronize_mesh_size(data_filter.n_nodes(), data_filter.n_cells(), comm);
```

bangerth replied:
My preference would be to actually spell out the name of the type, rather than use `auto`. When the `mesh_sizes` variable is used below and I want to understand what members it has, I need to look here to see what type it has, but then I need to go elsewhere (to the definition of `synchronize_mesh_size()` to *actually* see what the type is.

dabele replied:
Happy to follow your style.

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8851,7 +8896,553 @@ DataOutBase::write_hdf5_parallel(
+     */
+    NcFile() = default;
+
+  public:
+    /**
+     * Create a NetCDF file.
+     */
+    static NcFile
+    create(const std::string &filename, int open_mode, MPI_Comm comm)
+    {
+      NcFile ncfile;
+      auto   ncerr = nc_create_par(
+        filename.c_str(), open_mode, comm, MPI_INFO_NULL, &ncfile.ncid);
+      AssertThrowNC(ncerr);
+      return ncfile;
+    }
+
+    /**
+     * Open an existing NetCDF file for reading and/or writing.
+     */
+    static NcFile
+    open(const std::string &filename, int open_mode, MPI_Comm comm)
+    {
+      NcFile ncfile;
+      auto   ncerr = nc_open_par(
+        filename.c_str(), open_mode, comm, MPI_INFO_NULL, &ncfile.ncid);
+      AssertThrowNC(ncerr);
+      return ncfile;
+    }
```
You need these two `static` functions because you made the constructor `private`, but in truth, they are constructors of this class. I think you went this way because you didn't know how to write two constructors that take exactly the same set of arguments and you needed a way to distinguish. But how about a constructor like
```suggestion
    enum OpenOrCreate { open_file, create_file };
    
    NcFile (const std::string &filename, const int open_mode, const OpenOrCreate open_or_create, const  MPI_Comm comm)
    {
      switch (open_or_create) {
        case open_file: 
        {
          auto   ncerr = nc_create_par(
            filename.c_str(), open_mode, comm, MPI_INFO_NULL, &ncid);
          AssertThrowNC(ncerr);
       } 
       
       case create_file:
       { ... }
       
       default: DEAL_II_ASSERT_UNREACHABLE();
    }
```

Separately, `open_mode` and `comm` are input arguments and can be made `const`.

dabele replied:
const input arguments done. 

Constructors have some advantages over factory functions, being language constructs. But I don't agree that this is better here. My implementation is closer to the netcdf library. And function names are clearer than enum arguments to distinguish the operations, and there already is another enum argument (`open_mode`). This is internal utility, so not sure it's worth too much debate, if you prefer the constructor, I'll do that.

bangerth replied:
Since this is an internal class, as you correctly mention, it's fine if we leave it that way.

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8851,7 +8896,553 @@ DataOutBase::write_hdf5_parallel(
+   */
+  int
+  nc_put_att_var(int                                         ncid,
+                 int                                         varid,
```
Make all input arguments `const`, here and elsewhere.

dabele replied:
Done

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8851,7 +8896,553 @@ DataOutBase::write_hdf5_parallel(
+          {
+            return nc_put_att_text(ncid, varid, name, v.size(), v.c_str());
+          }
+        else if constexpr (std::is_same_v<T, double>)
+          {
+            return nc_put_att_double(ncid, varid, name, NC_DOUBLE, 1, &v);
+          }
+        else if constexpr (std::is_same_v<T, int>)
+          {
+            return nc_put_att_int(ncid, varid, name, NC_INT, 1, &v);
+          }
+      },
```
```suggestion
        else if constexpr (std::is_same_v<T, int>)
          {
            return nc_put_att_int(ncid, varid, name, NC_INT, 1, &v);
          }
          else
            DEAL_II_ASSERT_UNREACHABLE();
      },
```

dabele replied:
Done

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8851,7 +8896,553 @@ DataOutBase::write_hdf5_parallel(
+    // actually requires MPI communication to be certain
+    Assert(std::all_of(patches.begin(),
+                       patches.end(),
+                       [ref = patches[0].reference_cell](auto &&patch) {
```
```suggestion
                       [ref = patches[0].reference_cell](const auto &&patch) {
```

bangerth replied:
You'll also have to have an assert that `patches.size() > 0`. People call these functions with zero cells sometimes and then you'd get a segfault here :-)

dabele replied:
this is already checked in `write_cf_parallel`. But to be safe and make that clear, I added `Assert` with `ExcInternalError`.

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8851,7 +8896,553 @@ DataOutBase::write_hdf5_parallel(
+        y[i] = coords[i * spacedim + 1];
+      }
+    size_t coord_write_start[] = {(size_t)mesh_sizes.offset_nodes};
+    size_t coord_write_count[] = {(size_t)data_filter.n_nodes()};
```
These are both arrays of size 1, right? Perhaps make that clear by explicitly saying `[1]`.

dabele replied:
Done, all static arrays are sized now. Also made everything const that can be.

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8851,7 +8896,553 @@ DataOutBase::write_hdf5_parallel(
+    size_t cell_write_start[] = {(size_t)mesh_sizes.offset_cells,
+                                 size_t(0)};
+    size_t cell_write_count[] = {(size_t)data_filter.n_cells(),
+                                 (size_t)n_nodes_per_cell};
```
Same here with sizes. I think you can probably also make these variables `const`.

dabele replied:
Done

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8851,7 +8896,553 @@ DataOutBase::write_hdf5_parallel(
+              {
+                ncerr =
+                  nc_put_att_var(ncid, var_time, att_name.c_str(), att_value);
+                AssertThrowNC(ncerr);
+              }
+          }
+      }
+
+    // write the coordinate values
+    std::vector<double> coords;
+    data_filter.fill_node_data(coords);
+    std::vector<double> x(data_filter.n_nodes());
+    std::vector<double> y(data_filter.n_nodes());
+    for (unsigned int i = 0; i < mesh_sizes.num_nodes_local; ++i)
+      {
+        x[i] = coords[i * spacedim];
+        y[i] = coords[i * spacedim + 1];
+      }
+    size_t coord_write_start[] = {(size_t)mesh_sizes.offset_nodes};
+    size_t coord_write_count[] = {(size_t)data_filter.n_nodes()};
+    ncerr                      = nc_put_vara_double(
+      ncid, var_x, coord_write_start, coord_write_count, x.data());
+    AssertThrowNC(ncerr);
+    ncerr = nc_put_vara_double(
+      ncid, var_y, coord_write_start, coord_write_count, y.data());
+    AssertThrowNC(ncerr);
+
+    // write the cell node indices
+    std::vector<unsigned int> ucells;
+    data_filter.fill_cell_data(mesh_sizes.offset_nodes, ucells);
+    size_t cell_write_start[] = {(size_t)mesh_sizes.offset_cells,
+                                 size_t(0)};
+    size_t cell_write_count[] = {(size_t)data_filter.n_cells(),
+                                 (size_t)n_nodes_per_cell};
+    ncerr                     = nc_put_vara_uint(
+      ncid, var_cell, cell_write_start, cell_write_count, ucells.data());
+    AssertThrowNC(ncerr);
+  }
+
+  void
```
```suggestion
  }



  void
```

dabele replied:
added 3 empty lines after functions and types

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8851,7 +8896,553 @@ DataOutBase::write_hdf5_parallel(
+              ExcIO("You are trying to add a time point to an existing file "
+                    "that does not have a time dimension or variable. "
+                    "Check the options in DataOutBase::cf_flags."));
+          }
```
I think you can write this more easily as follows:
```suggestion
            AssertThrow(
              (inq_dim_time != NC_EBADDIM && inq_var_time != NC_ENOTVAR),
              ExcIO("You are trying to add a time point to an existing file "
                    "that does not have a time dimension or variable. "
                    "Check the options in DataOutBase::cf_flags."));
```
This way, the actual condition that is being checked shows up in the error message.

dabele replied:
I think I tried to keep netcdf error codes out of user visible error messages, they are not helpful. But I get your point, I think a good compromise is to introduce bool variables with informative names (e.g. `has_time_dimension`) and put those in the `Assert`.

bangerth replied:
Yes, that'd be great.

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8851,7 +8896,553 @@ DataOutBase::write_hdf5_parallel(
+        ncerr      = nc_get_var_double(ncid, var_time, times.data());
+        AssertThrowNC(ncerr);
+        auto it_lb = std::lower_bound(times.begin(), times.end(), flags.time);
+        index_t    = std::distance(times.begin(), it_lb);
```
If I see this right, you're only using `index_t` in this one place. Move the declaration from above to right here to minimize the lifetime of the variable to where it is actually used.

dabele replied:
it's used again lower down when writing the data sets. The uses are a bit too far apart for my taste, but this way it's only computed once. The whole function is quite long, I might try to split it up a bit.

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8851,7 +8896,553 @@ DataOutBase::write_hdf5_parallel(
+        AssertThrowNC(ncerr);
+        auto it_lb = std::lower_bound(times.begin(), times.end(), flags.time);
+        index_t    = std::distance(times.begin(), it_lb);
+        size_t count_t = 1;
```
This variable is only used several lines further down. Let's move the declaration of the variable right above the one call that uses it.

dabele replied:
moved.

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8851,7 +8896,553 @@ DataOutBase::write_hdf5_parallel(
+        AssertThrowNC(ncerr);
+#  endif
+      }
+  }
+} // namespace
+#endif // DEAL_II_WITH_NETCDF
+
+template <int dim, int spacedim>
```
Here and elsewhere below, let's have our usual three empty lines between functions.

dabele replied:
done for all functions.

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8851,7 +8896,553 @@ DataOutBase::write_hdf5_parallel(
+                               const std::string                &filename,
+                               const MPI_Comm                    comm)
+{
+  AssertThrow(spacedim == 2,
+              ExcMessage(
+                "DataOutBase can only write CF output in 2 dimensions."));
```
Oh, that's a bummer. Is that a limitation only of your current implementation, or does CF simply not have 3d formats? If the latter, I think that would be useful to state in the documentation of the functions you add to `DataOutInterface`.

dabele replied:
complicated answer: CF does have arbitrary dimensional variables, but only for structured grids. UGRID, the extension to CF for unstructured grids does have 3D (or 1D), but 3D is not quite mature as far as I can tell. UGRID gets slowly incorporated into CF, but only 2D is in so far, and 3D may be changed when integrating it. Also, it's quite trivial to go from 2D to 3D since there are more differences in the data format than maybe expected. For climate data, 2D is the most important.

For these reasons, I would do 3D (and maybe 1D) as a future addition. In the meantime, I will put it in the documentation.

bangerth replied:
OK, great, thank you for the explanation!

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8851,7 +8896,553 @@ DataOutBase::write_hdf5_parallel(
+  Assert((patches.size() > 0) || (n_ranks > 1), ExcNoPatches());
+
+  // The Netcdf routines perform a bunch of collective calls that expect all
+  // ranks to participate. One ranks without any patches we are missing
```
```suggestion
  // ranks to participate. On ranks without any patches we are missing
```

dabele replied:
Done

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8851,7 +8896,553 @@ DataOutBase::write_hdf5_parallel(
+    const int ierr  = MPI_Comm_split(comm, color, key, &split_comm);
+    AssertThrowMPI(ierr);
+  }
+  ScopeExit cleanup_on_exit{[&split_comm] {
+    int ierr = MPI_Comm_free(&split_comm);
+    AssertThrowMPI(ierr);
+  }};
```
Nice use of the `ScopeExit` class!

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -8851,7 +8896,553 @@ DataOutBase::write_hdf5_parallel(
+          // Explicitly checking that the mesh in the file is correct
+          // could be expensive. But NetCDF will complain if the dimensions
+          // don't match. So we only check that the file exists.
+          if (!std::filesystem::is_regular_file(filename))
+            {
+              AssertThrow(false,
+                          ExcIO(
+                            "You are trying to write data to an existing file "
+                            "but no file of the given name exists. "
+                            "Create the file with a mesh by setting "
+                            "CFFlags::keep_existing_file = false."));
+            }
+        }
```
```suggestion
          // Existing file with mesh definition will be reused.
          // Explicitly checking that the mesh in the file is correct
          // could be expensive. But NetCDF will complain if the dimensions
          // don't match. So we only check that the file exists.
              AssertThrow(std::filesystem::is_regular_file(filename),
                          ExcIO(
                            "You are trying to write data to an existing file "
                            "but no file of the given name exists. "
                            "Create the file with a mesh by setting "
                            "CFFlags::keep_existing_file = false."));
        }
```


Would it be useful to only run this check on the root process? If so, then perhaps something like this:
```suggestion
          // Existing file with mesh definition will be reused.
          // Explicitly checking that the mesh in the file is correct
          // could be expensive. But NetCDF will complain if the dimensions
          // don't match. So we only check that the file exists.
              AssertThrow(Utilities::MPI::broadcast(Utilities::MPI::this_mpi_process(comm) == 0 ?
                                                             std::filesystem::is_regular_file(filename)
                                                             : /* unimportant value */ false,
                                                             comm) == true,
                          ExcIO(
                            "You are trying to write data to an existing file "
                            "but no file of the given name exists. "
                            "Create the file with a mesh by setting "
                            "CFFlags::keep_existing_file = false."));
        }
```

dabele replied:
Changed to the first suggestion. Not sure the second is worth the complication.

bangerth reviewed (commented):
A whole bunch of comments, but nothing of great complexity. I think we can still get this into the upcoming release.

Would you please also add an entry into `doc/news/changes/minor` so you get credit where credit is due?

bangerth commented:
Oh, I see you already have a changelog entry. Thank you!

dabele commented:
fixed or answered all the comments I think.

still working on CI, something about cmake configuration of netcdf is not quite right.

bangerth commented:
I appreciate your work on this! Let me know once you think it's ready and you want me to take another look!

dabele commented:
- added some documentation and checks based on the discussion here
- fixed CI: unfortunately, ubuntu/apt libnetcdf packages contain broken cmake configs that point to the wrong include directory. So I removed the `find_package(CONFIG)` call. We already had a fallback to look for headers/libraries manually, so that fallback is now the only way to find netcdf. And I refined the feature check a little bit (checking for `parallel with netcdf-4` instead of just `parallel` which could (in theory, unlikely in practice) be insufficient.
- added a test using simplex meshes
- introduced a few utility functions to avoid repetition and make the functions shorter

@bangerth I think this is "done". Please have a look.

bangerth commented:
/rebuild

bangerth reviewed (approved):
Cool, thank you!

bangerth commented:
I believe that @drwells 's request has also been addressed, so I'm dismissing his hold on this PR.
